### PR TITLE
[Feature Extraction] Group 1 Feature

### PR DESCRIPTION
**Pull request by Kathryn-cat**
Added group 1 feature for block-level feature extraction.

We will test out the correctness of the extracted features in the future.

### CODE REVIEW

File: src/meta_schedule/feature_extractor/utils.h [hunk 1]
```diff
@@ -37,6 +37,10 @@ using MultiIndex = std::vector<PrimExpr>;
 using IntVec = std::vector<int64_t>;
 /*! \brief Vector of for loops */
 using ForVec = std::vector<const ForNode*>;
+/*! \brief Vector of blocks */
+using BlockVec = std::vector<const BlockRealizeNode*>;
+/*! \brief Vector of DFS paths */
+using PathVec = std::vector<const StmtNode*>;
```
Comment (junrushao) [inline src/meta_schedule/feature_extractor/utils.h:43]: nit: There is no need to name an alias if it's used not quite often :-)

File: src/meta_schedule/feature_extractor/per_block_feature.cc [hunk 2]
```diff
@@ -314,25 +311,109 @@ class PerBlockFeatureCollector : private StmtVisitor {
...
+  }
+
   void VisitStmt_(const ForNode* loop) final {
     int64_t auto_unroll;
+    const int64_t* extent = GetLoopIntExtent(loop);
     ForVec* for_vec = loop_nest_.Push(loop, &auto_unroll);
+    if ((extent && (*extent != 1)) || for_vec) {
+      outer_loop_prod_ *= *extent;
+      dfs_path_.push_back(loop);
+      analyzer_.Bind(loop->loop_var, loop->min);
```
Comment (junrushao) [inline src/meta_schedule/feature_extractor/per_block_feature.cc:342]: It's uncertain for me if we really want to do value binding here.

Some background:
- The analyzer is a tool that does integer set analysis, which provides basic functionality like proving whether or not two expressions are equal
- `Bind` means letting the analyzer know that a variable's value is always a constant - in our particular case, enforcing the assumption that `loop->loop_var == loop->min`

However, when the loop extent is not 1, for example,
```python
for i in range(0, 10): ... # extent is 10
```
binding `loop->loop_var` (i.e. `i`) to `loop->min` (i.e. `0`) doesn't seem to be a reasonable assumption.

Therefore, perhaps we should postpone this binding to follow-up PRs once we have better understanding of what this binding is actually used for.

File: src/meta_schedule/feature_extractor/per_block_feature.cc [hunk 3]
```diff
@@ -314,25 +311,109 @@ class PerBlockFeatureCollector : private StmtVisitor {
...
+    TVM_FEATURE_MATH_OP_ADD(int_other_func);
+    TVM_FEATURE_MATH_OP_ADD(bool_op);
+    TVM_FEATURE_MATH_OP_ADD(select_op);
+#undef TVM_FEATURE_MATH_OP_ADD
   }
 
   bool is_gpu_;
   int64_t cache_line_bytes_;
   int64_t arith_intensity_curve_num_samples_;
+  int64_t outer_loop_prod_ = 1;
```
Comment (junrushao) [inline src/meta_schedule/feature_extractor/per_block_feature.cc:400]: QQ: is this dup wrt `loop_nest_.prod`?

File: src/meta_schedule/feature_extractor/per_block_feature.cc [hunk 4]
```diff
@@ -314,25 +311,109 @@ class PerBlockFeatureCollector : private StmtVisitor {
...
   arith::Analyzer analyzer_;
+  BlockVec scopes_;
+  BlockVec ordered_blocks_;
```
Comment (junrushao) [inline src/meta_schedule/feature_extractor/per_block_feature.cc:403]: it doesn't seem to be used. Shall we remove it?

Comment (Kathryn-cat) [reply src/meta_schedule/feature_extractor/per_block_feature.cc:403]: They'll be used in other groups..?

File: src/meta_schedule/feature_extractor/per_block_feature.cc [hunk 5]
```diff
@@ -314,25 +311,109 @@ class PerBlockFeatureCollector : private StmtVisitor {
...
+    for (auto iter = dfs_path_.rbegin(); iter != dfs_path_.rend(); ++iter) {
+      const tir::StmtNode* stmt = *iter;
+      if (stmt == scope) {
+        break;
+      }
+      ICHECK(stmt->IsInstance<tir::ForNode>());
+      const int64_t* extent = GetLoopIntExtent(static_cast<const tir::ForNode*>(stmt));
+      if (*extent >= 1) {
+        prod_loop_extent *= *extent;
+      }
```
Comment (junrushao) [inline src/meta_schedule/feature_extractor/per_block_feature.cc:371]: No need to check if the extent is `>= 1`. On the other hand, null pointer needs careful checking

```suggestion
      if (const int64_t* extent = GetLoopIntExtent(static_cast<const tir::ForNode*>(stmt))) {
        prod_loop_extent *= *extent;
      }
```

File: src/meta_schedule/feature_extractor/per_block_feature.cc (hunk 4 above)
Comment (junrushao) [reply src/meta_schedule/feature_extractor/per_block_feature.cc:403]: We may remove it in this PR if it's not used in group 1

File: src/meta_schedule/feature_extractor/per_block_feature.cc (hunk 3 above)
Comment (Kathryn-cat) [reply src/meta_schedule/feature_extractor/per_block_feature.cc:400]: I'm wondering too. In the comments of Line 330 in utils, it says `loop_nest_.prod` is the product of the extents of all the loops, while `outer_loop_prod` is the product of the extents of outer loops. I don't know if they're the same or not?

File: src/meta_schedule/feature_extractor/per_block_feature.cc (hunk 4 above)
Comment (Kathryn-cat) [reply src/meta_schedule/feature_extractor/per_block_feature.cc:403]: okie

**Review by junrushao — APPROVED** [review]
LGTM

File: src/meta_schedule/feature_extractor/utils.h (hunk 1 above)
Comment (Kathryn-cat) [reply src/meta_schedule/feature_extractor/utils.h:43]: Addressed :)

File: src/meta_schedule/feature_extractor/per_block_feature.cc (hunk 2 above)
Comment (Kathryn-cat) [reply src/meta_schedule/feature_extractor/per_block_feature.cc:342]: okay, I commented it out

File: src/meta_schedule/feature_extractor/per_block_feature.cc (hunk 3 above)
Comment (Kathryn-cat) [reply src/meta_schedule/feature_extractor/per_block_feature.cc:400]: Addressed :)

File: src/meta_schedule/feature_extractor/per_block_feature.cc (hunk 5 above)
Comment (Kathryn-cat) [reply src/meta_schedule/feature_extractor/per_block_feature.cc:371]: Addressed :)

**Pull request merged**